File: tkpdreputation/src/main/java/com/tokopedia/tkpd/tkpdreputation/review/shop/domain/ReviewShopUseCase.java
package com.tokopedia.tkpd.tkpdreputation.review.shop.domain;

import com.tokopedia.design.utils.StringUtils;
import com.tokopedia.tkpd.tkpdreputation.domain.interactor.GetLikeDislikeReviewUseCase;
import com.tokopedia.tkpd.tkpdreputation.domain.model.GetLikeDislikeReviewDomain;
import com.tokopedia.tkpd.tkpdreputation.domain.model.LikeDislikeListDomain;
import com.tokopedia.tkpd.tkpdreputation.inbox.data.repository.ReputationRepository;
import com.tokopedia.tkpd.tkpdreputation.review.product.data.model.reviewlist.DataResponseReviewProduct;
import com.tokopedia.tkpd.tkpdreputation.review.product.data.model.reviewlist.DataResponseReviewShop;
import com.tokopedia.tkpd.tkpdreputation.review.product.data.model.reviewlist.Review;
import com.tokopedia.usecase.RequestParams;
import com.tokopedia.usecase.UseCase;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import rx.Observable;
import rx.functions.Func1;

/**
 * Created by zulfikarrahman on 1/19/18.
 */

public class ReviewShopUseCase extends UseCase<DataResponseReviewShop> {

    public static final String SHOP_DOMAIN = "shop_domain";
    public static final String SHOP_ID = "shop_id";
    public static final String PAGE = "page";
    public static final String PER_PAGE = "per_page";
    public static final String DEFAULT_PER_PAGE = "10";
    private static final String USER_ID = "user_id";
    private ReputationRepository reputationRepository;
    private GetLikeDislikeReviewUseCase getLikeDislikeReviewUseCase;

    @Inject
    public ReviewShopUseCase(ReputationRepository reputationRepository, GetLikeDislikeReviewUseCase getLikeDislikeReviewUseCase) {
        this.reputationRepository = reputationRepository;
        this.getLikeDislikeReviewUseCase = getLikeDislikeReviewUseCase;
    }

    @Override
    public Observable<DataResponseReviewShop> createObservable(final RequestParams requestParams) {
        return reputationRepository.getReviewShopList(requestParams.getParamsAllValueInString())
                .flatMap(new Func1<DataResponseReviewShop, Observable<DataResponseReviewShop>>() {
                    @Override
                    public Observable<DataResponseReviewShop> call(final DataResponseReviewShop dataResponseReviewShop) {
                        if(dataResponseReviewShop.getList()!= null && dataResponseReviewShop.getList().size() > 0) {
                            return getLikeDislikeReviewUseCase.createObservable(
                                    GetLikeDislikeReviewUseCase.getParam(createReviewIds(dataResponseReviewShop), requestParams.getString(USER_ID, "")))
                                    .map(new Func1<GetLikeDislikeReviewDomain, DataResponseReviewShop>() {
                                        @Override
                                        public DataResponseReviewShop call(GetLikeDislikeReviewDomain getLikeDislikeReviewDomain) {
                                            return mapLikeModelToReviewModel(getLikeDislikeReviewDomain, dataResponseReviewShop);
                                        }
                                    });
                        }else{
                            return Observable.just(dataResponseReviewShop);
                        }
                    }
                });
    }

    private DataResponseReviewShop mapLikeModelToReviewModel(GetLikeDislikeReviewDomain getLikeDislikeReviewDomain,
                                                                DataResponseReviewShop dataResponseReviewShop) {
        for(Review review : dataResponseReviewShop.getList()) {
            for (LikeDislikeListDomain likeDislikeListDomain : getLikeDislikeReviewDomain.getList()) {
                if(likeDislikeListDomain.getReviewId() == review.getReviewId()){
                    review.setTotalLike(likeDislikeListDomain.getTotalLike());
                    review.setLikeStatus(likeDislikeListDomain.getLikeStatus());
                    break;
                }
            }
        }
        return dataResponseReviewShop;
    }

    private String createReviewIds(DataResponseReviewShop dataResponseReviewShop) {
        List<String> listIds = new ArrayList<>();
        for(Review review :dataResponseReviewShop.getList()){
            listIds.add(String.valueOf(review.getReviewId()));
        }
        return StringUtils.convertListToStringDelimiter(listIds,"~");
    }

    public RequestParams createRequestParams(String shopDomain, String shopId, String page, String userId){
        RequestParams requestParams = RequestParams.create();
        requestParams.putString(SHOP_DOMAIN, shopDomain);
        requestParams.putString(SHOP_ID, shopId);
        requestParams.putString(PAGE, page);
        requestParams.putString(PER_PAGE, DEFAULT_PER_PAGE);
        requestParams.putString(USER_ID, userId);
        return requestParams;
    }
}

File: features/merchant/shop_page/src/main/java/com/tokopedia/shop/review/shop/domain/ReviewShopUseCase.kt
package com.tokopedia.shop.review.shop.domain

import com.tokopedia.shop.review.product.data.model.reviewlist.DataResponseReviewShop
import com.tokopedia.shop.review.shop.domain.model.GetLikeDislikeReviewDomain
import com.tokopedia.shop.review.shop.domain.repository.ReputationRepository
import com.tokopedia.usecase.RequestParams
import com.tokopedia.usecase.UseCase
import rx.Observable
import rx.functions.Func1
import java.util.*
import javax.inject.Inject

/**
 * Created by zulfikarrahman on 1/19/18.
 */
class ReviewShopUseCase @Inject constructor(private val reputationRepository: ReputationRepository, private val getLikeDislikeReviewUseCase: GetLikeDislikeReviewUseCase) : UseCase<DataResponseReviewShop>() {
    override fun createObservable(requestParams: RequestParams): Observable<DataResponseReviewShop> {
        return reputationRepository.getReviewShopList(requestParams.paramsAllValueInString)
                .flatMap(Func1<DataResponseReviewShop?, Observable<DataResponseReviewShop>> { dataResponseReviewShop: DataResponseReviewShop? ->
                    if (dataResponseReviewShop!!.list != null && dataResponseReviewShop.list!!.size > 0) {
                        return@Func1 getLikeDislikeReviewUseCase.createObservable(
                                GetLikeDislikeReviewUseCase.Companion.getParam(createReviewIds(dataResponseReviewShop), requestParams.getString(USER_ID, "")))
                                .map { getLikeDislikeReviewDomain: GetLikeDislikeReviewDomain? -> mapLikeModelToReviewModel(getLikeDislikeReviewDomain, dataResponseReviewShop) }
                    } else {
                        return@Func1 Observable.just(dataResponseReviewShop)
                    }
                })
    }

    private fun mapLikeModelToReviewModel(getLikeDislikeReviewDomain: GetLikeDislikeReviewDomain?,
                                          dataResponseReviewShop: DataResponseReviewShop?): DataResponseReviewShop? {
        for (review in dataResponseReviewShop!!.list!!) {
            for (likeDislikeListDomain in getLikeDislikeReviewDomain!!.list!!) {
                if (likeDislikeListDomain.reviewId == review.reviewId) {
                    review.totalLike = likeDislikeListDomain.totalLike
                    review.likeStatus = likeDislikeListDomain.likeStatus
                    break
                }
            }
        }
        return dataResponseReviewShop
    }

    private fun createReviewIds(dataResponseReviewShop: DataResponseReviewShop?): String {
        val listIds: MutableList<String> = ArrayList()
        for (review in dataResponseReviewShop!!.list!!) {
            listIds.add(java.lang.String.valueOf(review.reviewId))
        }
        return listIds.joinToString("~")
    }

    fun createRequestParams(shopDomain: String?, shopId: String?, page: String?, userId: String?): RequestParams {
        val requestParams = RequestParams.create()
        requestParams.putString(SHOP_DOMAIN, shopDomain)
        requestParams.putString(SHOP_ID, shopId)
        requestParams.putString(PAGE, page)
        requestParams.putString(PER_PAGE, DEFAULT_PER_PAGE)
        requestParams.putString(USER_ID, userId)
        return requestParams
    }

    companion object {
        const val SHOP_DOMAIN = "shop_domain"
        const val SHOP_ID = "shop_id"
        const val PAGE = "page"
        const val PER_PAGE = "per_page"
        const val DEFAULT_PER_PAGE = "12"
        private const val USER_ID = "user_id"
    }

}